syntax = "proto3";
package gateway;
import "celaut.proto";
import "buffer.proto";

message GasAmount {
    string n = 1;
}

message TokenMessage {
    string token = 1;
}

message Refund {
    GasAmount amount = 1;
}

message Payment {
    string deposit_token = 1;
    celaut.Service.Api.ContractLedger contract_ledger = 3;
    GasAmount gas_amount = 4;
}

message Metrics {
    GasAmount gas_amount = 1;
}

message Instance {
    optional celaut.Any.Metadata instance_meta = 1;
    celaut.Instance instance = 2;
    optional string token = 3;
}

message Client {
    string client_id = 1;
}

message RecursionGuard {
    string token = 1;
}

message ServiceWithMeta {
    celaut.Any.Metadata metadata = 1;
    celaut.Service service = 2;
}

// ( celaut.Any.Metadata.HashTag.Hash=H, ServiceWithMeta=S, celaut.Configuration=C; { H v S v H^C v S^C } )
message HashWithConfig { // H^C
    celaut.Any.Metadata.HashTag.Hash hash = 1;
    celaut.Configuration config = 3;
    optional celaut.Sysresources min_sysreq = 4;  // init system specs.
    optional celaut.Sysresources max_sysreq = 5;  // minimun system specs that the node needs to 've for it sometime.
    optional GasAmount initial_gas_amount = 6;
}
message ServiceWithConfig { // S^C
    ServiceWithMeta service = 2;
    celaut.Configuration config = 3;
    optional celaut.Sysresources min_sysreq = 4;  // init system specs.
    optional celaut.Sysresources max_sysreq = 5;  // minimun system specs that the node needs to 've for it sometime.
    optional GasAmount initial_gas_amount = 6;
}

message ModifyServiceSystemResourcesOutput {
    celaut.Sysresources sysreq = 1;
    GasAmount gas = 2;
}

message ModifyServiceSystemResourcesInput {
    celaut.Sysresources min_sysreq = 1;
    celaut.Sysresources max_sysreq = 2;
}

// GRPC.

service Gateway {
    rpc StartService( stream buffer.Buffer) returns ( stream buffer.Buffer) {} // (Client, celaut.Any.Metadata.HashTag.Hash=H, celaut.Any=S, celaut.Configuration & celaut.Sysresources=C; { H v S v H^C v S^C } ) -> Instance
    rpc StopService(stream buffer.Buffer) returns (stream buffer.Buffer) {} // TokenMessage -> Refund

    rpc GetInstance(stream buffer.Buffer) returns (stream buffer.Buffer) {} //  -> Instance
    rpc GenerateClient(stream buffer.Buffer) returns (stream buffer.Buffer) {} //  -> Client

    rpc ModifyServiceSystemResources(stream buffer.Buffer) returns (stream buffer.Buffer) {} // ModifyServiceSystemResourcesInput ->

    rpc GetMetrics(stream buffer.Buffer) returns (stream buffer.Buffer) {} // TokenMessage -> Metrics
}